### PR TITLE
ARROW-5027: [Python] Python bindings for JSON reader

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR implements Python bindings for the JSON reader.

### CODE REVIEW

**Comment by wesm** [comment]
@pitrou would you like to take a turn on the JSON reader C++ code since it's fairly related to CSV reading? We eventually need to place a common Dataset interface in front of both of these (and Parquet, and ORC, too) so that they can be consumed in Scan nodes in AQE (query engine)

**Comment by pitrou** [comment]
@wesm No problem. What is the exact status right now?

**Comment by wesm** [comment]
That would be a question for @bkietz. My understanding is that multi-threaded parsing and conversion is not yet implemented, and some aspects of type inference may also be incomplete.

Note that with all file formats there are a couple of distinct modes of operation that we have to support:

* Parsing and conversion where the schema is unknown or partially known (e.g. only a few column types). In the case of CSV and JSON, it must be inferred. In the case of Parquet, it is found in the file footer
* Parsing and conversion for a known and fixed schema or a partially known schema. The schema may have been determined and stored as part of a prior inference step. This is the traditional mode of execution for analytic SQL systems. A typical case is for a user to write down the schema they want to apply to a file, with some conformance rules (like whether to error on unsafe value casts or not, whether certain fields are allowed to contain nulls or not)

**Comment by bkietz** [comment]
Currently, JSON can only be read in a single thread from a single buffer. Type inference is implemented but can't currently use a partial schema. Providing a full schema for conversion is supported.

I'm working implementing the full multithreaded `TableReader` this week. For the first draft type inference will be applied to the first parsed record batch, then that schema will be locked in for subsequent record batches.

**Comment by wesm** [comment]
Ok, @bkietz and @pitrou I'll leave it to you two to coordinate the JSON development tasks but I would like to see this shipped in a usable state for query processing in 0.14 (i.e. over the next 8 weeks or less)

**Comment by pitrou** [comment]
> I'm working implementing the full multithreaded TableReader this week. For the first draft type inference will be applied to the first parsed record batch, then that schema will be locked in for subsequent record batches.

Sounds fine to me.

**Comment by nealrichardson** [comment]
FYI https://github.com/apache/arrow/pull/4165 has merged, so you should be able to proceed with this now.

**Comment by jorisvandenbossche** [comment]
@pcmoritz do you have time to update this the coming days? (otherwise someone else could take over if you want)

**Comment by pcmoritz** [comment]
Thanks, I'm looking into getting this merged.

The current blocker is that the json.py file is clashing with the standard json package. Are there any preferences on how to go ahead with this? Choices I can think of right now include:

- Add `from __future__ import absolute_import` (it should probably be added everywhere, but at least in files where the standard json module is used)
- Rename pyarrow.json to something else
- Maybe more pyarrow.json and pyarrow.csv into a common namespace, potentially even into the top-level pyarrow namespace?

**Comment by pitrou** [comment]
@pcmoritz I'd like to review this before it gets merged. Probably on Monday.

As for the name clash issue, I think we should go with `from __future__ import absolute_import`.

File: python/pyarrow/_json.pyx
Comment (pitrou) [inline python/pyarrow/_json.pyx]: `convert_options` is unused here.

File: python/pyarrow/_json.pyx
Comment (pitrou) [inline python/pyarrow/_json.pyx]: A docstrings is missing here.

File: python/pyarrow/_json.pyx
Comment (pitrou) [reply python/pyarrow/_json.pyx]: Also, can you add a docstring?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this! It looks like this is missing some tests, though.

File: python/pyarrow/_json.pyx
Comment (pcmoritz) [reply python/pyarrow/_json.pyx]: Both fixed, thanks!

File: python/pyarrow/_json.pyx
Comment (pcmoritz) [reply python/pyarrow/_json.pyx]: fixed

**Comment by pcmoritz** [comment]
This is ready to merge now

File: python/pyarrow/_json.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_json.pyx]: Mention the default here?

**Review by jorisvandenbossche — COMMENTED** [review]
Minor doc comment, but I see this was copied as is from the csv docs, so not that important.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
The MinGW failure looks unrelated.

**Pull request closed without merging**
